### PR TITLE
Use WHERE rather than HAVING clause in group ordering TAQL

### LINKED ISSUE #94: Using columns in both group_cols and TaQL returns empty dataset
* dask-ms version: 0.2.3
* Python version: 3.7.4
* Operating System: Ubuntu 16.04.6 LTS (Xenial Xerus)

I'm attempting to use xds_from_ms to select corrected data from a specific baseline in a dataset. Using the query structure:

```
mytaql = '(FIELD_ID==0) && (DATA_DESC_ID==0) && (ANTENNA1==2) && (ANTENNA2==5)'

corrected_data = daskms.xds_from_ms(myms, columns=['CORRECTED_DATA', 'TIME', 'FLAG', 'FIELD_ID', 'UVW'], taql_where=mytaql)
```
Returns an empty dataset, whereas the same query run in the table environment of CASA returns 300 rows. It's clear that the data is there, but somehow not picked up. If I drop the antenna requirements, the expected xarray dataset is returned.

Cheers

### LINKED ISSUE #100: Exact logical opposite of #94!
* dask-ms version: pip
* Python version: 3.6
* Operating System: 18.04

### Description

I see the exact logical opposite of #94! Actually @IanHeywood is saying the same in #98, but I thought I'd escalate that to an issue.

If the taql includes antennas, ANTENNA1 **must** be in group_cols, otherwise ``daskms.xds_from_ms()`` returns an empty iterator. 

If taql doesn't include antennas, then the iterator works as expected even without ANTENNA1 in group_col.

### What I Did

My branch of shadeMS. @sjperkins let me know if you want to play with it directly. But basically:

```
2020-04-15 17:45:53 - shadems - INFO - ------------------------------------------------------
2020-04-15 17:45:53 - shadems - DEBUG - taql is ((ANTENNA1==6 || ANTENNA2==6) || (ANTENNA1==10 || ANTENNA2==10)) && (FIELD_ID==1), group_cols is ['FIELD_ID', 'DATA_DESC_ID']
2020-04-15 17:45:53 - shadems - INFO -                  : Reading MS, please wait
2020-04-15 17:45:53 - shadems - INFO -                  : rendering 0 points
```

0 points means the iterator returned nothing. Same run with ANTENNA1 in group_cols:
```
2020-04-15 17:45:15 - shadems - DEBUG - taql is ((ANTENNA1==6 || ANTENNA2==6) || (ANTENNA1==10 || ANTENNA2==10)) && (FIELD_ID==1), group_cols is ['FIELD_ID', 'DATA_DESC_ID', 'ANTENNA1']
2020-04-15 17:45:15 - shadems - INFO -                  : Reading MS, please wait
2020-04-15 17:45:15 - shadems - DEBUG - axis Frequency has shape (1, 64)
2020-04-15 17:45:15 - shadems - DEBUG - axis Phase has shape (1387, 64)
2020-04-15 17:45:15 - shadems - DEBUG - axis Frequency has shape (1, 64)
2020-04-15 17:45:15 - shadems - DEBUG - axis Phase has shape (1095, 64)
2020-04-15 17:45:15 - shadems - INFO -                  : rendering 1.59e+05 points
```

### PR DESCRIPTION
**Pull request by sjperkins**
Fixes #94 
Fixes #100

- [x] Tests added / passed

  ```bash
  $ py.test -v -s daskms/tests
  ```

  If the pep8 tests fail, the quickest way to correct
  this is to run `autopep8` and then `flake8` and
  `pycodestyle` to fix the remaining issues.

  ```
  $ pip install -U autopep8 flake8 pycodestyle
  $ autopep8 -r -i daskms
  $ flake8 daskms
  $ pycodestyle daskms
  ```

- [x] Fully documented, including `HISTORY.rst` for all changes
      and one of the `docs/*-api.rst` files for new API

  To build the docs locally:

  ```
  pip install -r requirements.readthedocs.txt
  cd docs
  READTHEDOCS=True make html
  ```

### CODE REVIEW

**Comment by sjperkins** [comment]
@IanHeywood Would it be possible for you to provide general feedback?

Essentially dask-ms will warn about unexpected behaviour if a column is present in both group_cols and taql_where, but not actively fail. I could change it to actively fail, and that might actually be a better scenario.

**Comment by sjperkins** [comment]
On seconds thoughts I'm going to make it fail instead. I'm not a fan of users discovering unexpected behaviour after the fact.

**Comment by o-smirnov** [comment]
Hmmm, but what if I use taql to select a subset of antennas, and yet want to group the result by antenna? Surely that's a legit use case? Or am I missing something?

**Comment by o-smirnov** [comment]
OK, having read the docs and the gitter discussion, let me see if I understand:

* ``dask-ms`` can group the data by specific columns, which is a very straightforward way to e.g. iterate over fields, or scans, or some such.

* @IanHeywood was trying to apply this to iterate over antennas. But this behaviour is not natively supported by ``group_col``, because it's a non-trivial combination of two columns that the data would need to be grouped by. And, in any case, I think the ``group_cols`` feature sort of implies separating the MS into non-overlapping subsets, whereas the subset of data per each antenna overlaps with other such subsets.

* @sjperkins's snippet at https://gitter.im/dask-ms/community?at=5e95bf92c52b6a38abe67e75 shows how to accomplish this access pattern.

Bottom line is, by-antennas are not trivial, but manageable -- the snippet is short enough.

That all understood, I still don't understand the nature of the warning in this PR. Nevermind antennas, but what if I want a subset of scans (``SCAN_NUMBER<20`` in the taql), grouped by scan? Why should that be suspicious?

**Comment by IanHeywood** [comment]
> Essentially dask-ms will warn about unexpected behaviour if a column is present in both group_cols and taql_where, but not actively fail.

What I observed seemed to be that a column _had_ to be present in group_cols for TaQL queries to work, i.e. not return empty arrays. So failing if TaQL mentions a column that is also in group_cols would break TaQL functionality entirely.

**Comment by IanHeywood** [comment]
This demonstrates the issue:

https://gist.github.com/IanHeywood/5e8f7b739bde5286978610a799f77f83

**Comment by IanHeywood** [comment]
I only get anything back from the DATA column when the group_cols contains the columns that are mentioned in TaQL:

```
ANTENNA1 == 5 || ANTENNA2 == 5
['DATA_DESC_ID']

ANTENNA1 == 5 || ANTENNA2 == 5
['DATA_DESC_ID', 'ANTENNA1', 'ANTENNA2']
(970, 1, 4)
(970, 1, 4)
(971, 1, 4)
(975, 1, 4)
(969, 1, 4)
(967, 1, 4)
(975, 1, 4)
(968, 1, 4)
(972, 1, 4)
(974, 1, 4)
(972, 1, 4)
(975, 1, 4)
(975, 1, 4)
(975, 1, 4)
(975, 1, 4)
(971, 1, 4)
(969, 1, 4)
(971, 1, 4)
(947, 1, 4)
(974, 1, 4)
(971, 1, 4)
(970, 1, 4)
(974, 1, 4)
(974, 1, 4)
(970, 1, 4)
(974, 1, 4)
(970, 1, 4)
(970, 1, 4)
(971, 1, 4)
(975, 1, 4)
(969, 1, 4)
(967, 1, 4)
(975, 1, 4)
(968, 1, 4)
(972, 1, 4)
(974, 1, 4)
(972, 1, 4)
(975, 1, 4)
(975, 1, 4)
(975, 1, 4)
(975, 1, 4)
(971, 1, 4)
(969, 1, 4)
(971, 1, 4)
(947, 1, 4)
(974, 1, 4)
(971, 1, 4)
(970, 1, 4)
(974, 1, 4)
(974, 1, 4)
(970, 1, 4)
(974, 1, 4)

FIELD_ID==2
['DATA_DESC_ID', 'FIELD_ID']
(9596, 1, 4)
(9596, 1, 4)

FIELD_ID==2
['DATA_DESC_ID', 'ANTENNA1', 'ANTENNA2']
```

**Comment by sjperkins** [comment]
> I still don't understand the nature of the warning in this PR. Nevermind antennas, but what if I want a subset of scans (`SCAN_NUMBER<20` in the taql), grouped by scan? Why should that be suspicious?

It actually might not be -- there are test cases where this kind of thing [works](https://github.com/ska-sa/dask-ms/blob/45e9c1ad4490fb75b026a01cfe7bbaaa41c06152/daskms/tests/test_ms_read_and_update.py#L160-L165), admittedly without the group_cols interaction.

A potential problem with group_cols is that the related columns (SCAN_NUMBER in your case), are not exposed as dask arrays but as concrete integers on datasets. The thinking behind this is to allow joins on subtables (e.g. FIELD_ID references rows in the FIELD subtable, especially for accessing PHASE_DIR, DATA_DESC_ID references rows in the DATA_DESCRIPTION subtable for accessing the correct SPW).

There's certainly a sub-text here of me encouraging power users to re-implement TAQL-like functionality in numpy, dask and dask-ms. For example in the https://gitter.im/dask-ms/community?at=5e95bf92c52b6a38abe67e75. The technical reason for this is performance: dask needs reasonably big chunks of fixed shape data, loaded contiguously from disk to perform.

By contrast, TAQL is a broad and powerful language, that may introduce non-contiguous disk access. Also, it's grouping and ordering has not been exercised a lot (by me) for *applications* of dask-ms, so I'm not entirely sure how it's interacting the dask-ms's internals. https://github.com/ska-sa/dask-ms/issues/94 indicated to me that at least to the end user, unexpected behaviour is occurring. But, because CASA users are familiar with it, they're going to bring their paradigm to dask-ms.

My feeling on this is that either the user needs to be warned, or the problematic aspects of the functionality disabled (no columns in both group_cols and TAQL) to prevent unexpected behaviour.

**Comment by sjperkins** [comment]
> What I observed seemed to be that a column _had_ to be present in group_cols for TaQL queries to work, i.e. not return empty arrays. So failing if TaQL mentions a column that is also in group_cols would break TaQL functionality entirely.

> This demonstrates the issue:
> 
> https://gist.github.com/IanHeywood/5e8f7b739bde5286978610a799f77f83

Thanks, thats a neat example demonstrating what's happening and so close to a test case that I had to turn it into one in https://github.com/ska-sa/dask-ms/pull/98/commits/71f1a338ba36837e8d765b77e532247f9c4890e8 :-)

I  think what's happening here is that a request for only the `DATA` column is present in the `xds_from_ms` call.

```python
	msdata = xms.xds_from_ms(myms,
			columns = ['DATA'],
			group_cols = group_cols,
			taql_where = mytaql)
```

`columns = ['DATA']` so `ANTENNA1` and `ANTENNA2` aren't ending ending up on the dataset. I think it'll work if you add them in.

I still need to think if this is an acceptable interaction with the TAQL and whether further exceptions or warnings should be issued so that dask-ms isn't behaving unexpectedly.

I also want to say that this is useful information because unexpected behaviour is not desirable.

**Comment by sjperkins** [comment]
> I think what's happening here is that a request for only the DATA column is present in the xds_from_ms call.

I spoke to soon.

1. If ANTENNA1 and ANTENNA2 are in **group_cols**, they'll exist on the dataset as concrete integers. Mostly because if one is grouping there's an implied join with a subtable. The ANTENNA subtable join is probably not useful in this case, but joins with DATA_DESCRIPTION and FIELD_ID almost always are.
2. If ANTENNA1 and ANTENNA2 aren't in **group_cols** and are in **columns** then they'll be dask arrays.
3. If ANTENNA1 and ANTENNA2 aren't in **group_cols** and not in **columns** then they're simply not present.

I'd accept that laws of least astonishment are probably being violated here. As always a tension between what needs to be technically achievable and being confusing.

**Comment by sjperkins** [comment]
In fact (1) is not documented in the [API](https://dask-ms.readthedocs.io/en/latest/_modules/daskms/dask_ms.html#xds_from_table), but is in the [tutorial](https://dask-ms.readthedocs.io/en/latest/tutorial/reads.html#grouping) so it is certainly going to be somewhat astonishing!

**Comment by IanHeywood** [comment]
>  I think it'll work if you add them in.

It doesn't work if `columns` but not `group_cols` contains the columns that feature in the TaQL:

```
import daskms as xms

myms = 'AF0236_spw01.ms'

antenna_taql = 'ANTENNA1 == 5 || ANTENNA2 == 5'

default_group_cols = ['DATA_DESC_ID']

def get_data(myms,mytaql,group_cols):
	msdata = xms.xds_from_ms(myms,
			columns = ['DATA','ANTENNA1','ANTENNA2'],
			group_cols = group_cols,
			taql_where = mytaql)

	print(mytaql)
	print(group_cols)
	for group in msdata:
		print(group.DATA.shape)
	print('')

	return

get_data(myms,mytaql='',group_cols=default_group_cols)
get_data(myms,antenna_taql,group_cols=default_group_cols)
```

The second run gives the crash:

```
['DATA_DESC_ID']
(341677, 1, 4)
(341677, 1, 4)

Traceback (most recent call last):
  File "taql_demo.py", line 26, in <module>
    get_data(myms,antenna_taql,group_cols=default_group_cols)
  File "taql_demo.py", line 15, in get_data
    taql_where = mytaql)
  File "/home/ianh/.local/lib/python2.7/site-packages/daskms/dask_ms.py", line 291, in xds_from_ms
    **kwargs)
  File "/home/ianh/.local/lib/python2.7/site-packages/daskms/dask_ms.py", line 229, in xds_from_table
    **kwargs).datasets()
  File "/home/ianh/.local/lib/python2.7/site-packages/daskms/reads.py", line 395, in datasets
    self.index_cols, self.chunks)
  File "/home/ianh/.local/lib/python2.7/site-packages/daskms/ordering.py", line 196, in group_row_ordering
    for g, nrow in enumerate(nrows):
TypeError: 'NoneType' object is not iterable
```

**Comment by sjperkins** [comment]
> It doesn't work if columns but not group_cols contains the columns that feature in the TaQL:

OK, still working to try reproduce this on my small test dataset.

File: daskms/tests/test_github.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+    assert len(datasets) == 2
+    assert datasets[0].DATA_DESC_ID == 0
+    assert datasets[1].DATA_DESC_ID == 1
+
+    for ds in datasets:
+        expr = da.logical_or(ds.ANTENNA1.data == 5, ds.ANTENNA2.data == 5)
+        expr, equal = dask.compute(expr, da.all(expr))
+        assert equal.item() is True
+        assert len(expr) > 0
```
Comment (sjperkins) [inline daskms/tests/test_github.py:29]: @IanHeywood Thanks for providing me with the MS for which you were encountering this problem. This test case failed for me prior to this commit by simply returning an empty iterator (and I believe reproducing @o-smirnov 's #100).

With the change in this commit, it does return two datasets (different DDID's but probably same SCAN/FIELD as the rows are the same) whose ANTENNA1 and ANTENNA2 values are different.

Comment (sjperkins) [reply daskms/tests/test_github.py:29]: Does this change fix the problem for you on your side? I couldn't reproduce your crash in this branch.

**Comment by o-smirnov** [comment]
@sjperkins this fixes the problem. Let's merge!
